app/lib/account/backend.dart

```diff
@@ -0,0 +1,193 @@
+import 'package:googleapis/oauth2/v2.dart' as oauth2_v2;
+import 'package:http/http.dart' as http;
+import 'package:logging/logging.dart';
+import 'package:retry/retry.dart';
+import 'package:uuid/uuid.dart';
+
+import '../frontend/models.dart' show PackageVersion;
+
+import 'models.dart';
+
+final _logger = new Logger('pub.account.backend');
+final _retry = RetryOptions();
+final _uuid = Uuid();
+
+/// The pub client's OAuth2 identifier.
+final _pubAudience = '818368855108-8grd2eg9tj9f38os6f1urbcvsq399u8n.apps.'
+    'googleusercontent.com';
+
+/// Sets the account backend service.
+void registerAccountBackend(AccountBackend backend) =>
+    ss.register(#_accountBackend, backend);
+
+/// The active account backend service.
+AccountBackend get accountBackend =>
+    ss.lookup(#_accountBackend) as AccountBackend;
+
+/// Represents the backend for the account handling and authentication.
+class AccountBackend {
+  final DatastoreDB _db;
+  final _defaultAuthProvider = GoogleOauth2AuthProvider(_pubAudience);
+
+  AccountBackend(this._db);
+
+  Future close() async {
+    await _defaultAuthProvider.close();
+  }
+
+  /// Returns the `User` entry for the [userId] or null if it does not exists.
+  Future<User> lookupUserById(String userId) async {
+    final key = _db.emptyKey.append(User, id: userId);
+    return (await _db.lookup<User>([key])).single;
+  }
+
+  /// Returns the `User` entry that is associated with the [accessToken].
+  ///
+  /// When no entry exists in Datastore, this method will create a new one.
+  ///
+  /// When the authenticated e-mail of the user changes, the email field will
+  /// be updated to the latest one.
+  ///
+  /// The method returns null if the access token is invalid.
+  Future<User> authenticateWithAccessToken(String accessToken) async {
+    final auth = await _defaultAuthProvider.tryAuthenticate(accessToken);
+    if (auth == null) {
+      return null;
+    }
+    return await _lookupOrCreateUserByOauthUserId(auth);
+  }
+
+  Future<User> _lookupOrCreateUserByOauthUserId(AuthResult auth) async {
+    final mappingKey = _db.emptyKey.append(User, id: auth.userId);
+
+    final user = await _retry.retry(() async {
+      // Check existing mapping.
+      final mapping = (await _db.lookup<OAuthUserID>([mappingKey])).single;
+      if (mapping != null) {
+        final user = (await _db.lookup<User>([mapping.userIdKey])).single;
+        // TODO: we should probably have some kind of consistency mitigation
+        if (user == null) {
+          throw StateError('Incomplete OAuth userId mapping.');
+        }
+        return user;
+      }
+
+      // Query the first use of the e-mail address.
+      DateTime created = DateTime.now().toUtc();
+      final uploadedVersions = _db.query<PackageVersion>()
+        ..filter('uploaderEmail =', auth.email);
+      await for (var version in uploadedVersions.run()) {
+        if (created.isAfter(version.created)) {
+          created = version.created;
+        }
+      }
```

Comment:
I suggest we backfull the `User` entities in a one-off script...
This way, the `User` entity might exist with a `user.oauthUserId == null`, until the user logs in the first time... at which point we create the `OAuthUserID` entity AND update `user.oauthUserId = ...` in the same transaction...

This way, our one-off script can move us completely away from emails... We can replace the `packageVersion.uploaderEmail` with `packageVersion.uploader` and `package.uploaderEmails` with `package.uploaders`.
Then we can make the old fields as deprecated and as soon as we switch over traffic we stop using them.


Note: if we really smart, we design the one-off migration script such that it searches for a `User` entity with the given email before creating one. This way we can ensure that we can re-run the one-off script multiple times... Thus, we'll run it before and after we switch over traffic, how does that sound?

Reply:
PTAL: I've updated the lookup to handle a pre-migrated User entity too. (I'd do the migration script in a separate PR)